layabout.py

```diff
@@ -1,3 +1,287 @@
+        """
+        Connect to the Slack API and run the event handler loop.
+
+        Args:
+            token: A Slack API token. If absent an attempt will be made to use
+                the environment variable supplied at instantiation.
+            interval: The number of seconds to wait between fetching events
+                from the Slack API.
+            retries: The number of retry attempts to make if a connection to
+                Slack if not established or is lost.
+            backoff: The strategy used to determine how long to wait between
+                retries. Must take as input the number of the current retry and
+                output a :obj:`float`. If absent an exponential backoff
+                strategy will be used.
+            until: The condition used to evaluate whether this method
+                terminates. Must take as input an :obj:`list` of :obj:`dict`
+                representing Slack RTM API events and return a :obj:`bool`. If
+                absent this method will run forever.
+
+        Raises:
+            FailedConnection: If connecting to the Slack API fails.
+        """
+        backoff = backoff or _exponential
+        until = until or _forever
+
+        # The initial connection may use a given token or attempt to use an
+        # environment variable.
+        if not self.connect(token=token):
+            if not self._reconnect(retries=retries, backoff=backoff):
+                raise FailedConnection('Failed to connect to the Slack API')
+
+        # TODO: Should we force callers to handle KeyboardInterrupt on their
+        # own, or should we try to handle it for them? 🤔
+        while True:
+            try:
+                # Fetch new RTM events from the API.
+                events = self.slack.rtm_read()
+
+            # This is necessary to handle an error caused by a bug in Slack's
+            # Python client. For more information see
+            # https://github.com/slackhq/python-slackclient/issues/127
+            #
+            # TODO: The TimeoutError could be more elegantly resolved by making
+            # a PR to the websocket-client library and letting them coerce that
+            # exception to a WebSocketTimeoutException.
+            except (WebSocketConnectionClosedException, TimeoutError):
+                log.debug('Lost connection to the Slack API, attempting to '
+                          'reconnect')
+                # Attempt to re-use our existing SlackClient and token.
+                if self._reconnect(retries=retries, backoff=backoff):
+                    log.debug('Reconnected to the Slack API')
+                    # We continue so we can fetch new events before proceeding.
+                    continue
+
+                raise FailedConnection('Failed to reconnect to the Slack API')
+
+            if not until(events):
+                # TODO: Is this even a good debugging message?
+                log.debug('Terminal condition met')
```

Comment:
No. I have no idea what it's trying to say or why it's here.

Reply:
Eh, it's mostly here just to log why we've exited the Layabout event loop early. Imagine the following:
```python
>>> import logging
>>> from layabout import Layabout
>>> logging.basicConfig(level=logging.DEBUG)
>>> logger = logging.getLogger('app')
>>> layabout = Layabout('app')
>>> def not_even_once(events):
...     logger.debug("Found these events: %s", events)
...     return False
...
>>> layabout.run(until=not_even_once)
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): slack.com                      
DEBUG:urllib3.connectionpool:https://slack.com:443 "POST /api/rtm.start HTTP/1.1" 200 None     
DEBUG:app:Encountered these events: []         
DEBUG:layabout:Terminal condition met
>>>
```
Maybe it would be more helpful if it was
> Terminal condition met. Exiting event loop

or perhaps it's just so obvious that it's altogether unnecessary... :thinking:

Reply:
That makes much more sense. Even in the context of the other log lines the Terminal condition met phrase doens't make a whole lot of sense to.

I've also read a ton of logs so ...

Reply:
It's slightly better with "Exiting event loop". I think that statement alone is probably a better log message. If the user wants their terminal condition function to log something when it returns `False` they can do that themselves.